File: go/src/ipaddress/ipv6_mapped_test.go
package ipaddress

import "testing"
import "math/big"
import "fmt"

type IPv6MappedTest struct {
	ip                           IPAddress
	s                            string
	sstr                         string
	_str                         string
	u128                         big.Int
	address                      string
	valid_mapped                 map[string]big.Int
	valid_mapped_ipv6            map[string]big.Int
	valid_mapped_ipv6_conversion map[string]string
}

func ipv6MappedSetup() IPv6MappedTest {
	valid_mapped := map[string]big.Int{}
	valid_mapped["::13.1.68.3"] = str2Int("281470899930115", 10)
	valid_mapped["0:0:0:0:0:ffff:129.144.52.38"] = str2Int("281472855454758", 10)
	valid_mapped["::ffff:129.144.52.38"] = str2Int("281472855454758", 10)
	valid_mapped_ipv6 := map[string]big.Int{}
	valid_mapped_ipv6["::ffff:13.1.68.3"] = str2Int("281470899930115", 10)
	valid_mapped_ipv6["0:0:0:0:0:ffff:8190:3426"] = str2Int("281472855454758", 10)
	valid_mapped_ipv6["::ffff:8190:3426"] = str2Int("281472855454758", 10)
	valid_mapped_ipv6_conversion := map[string]string{}
	valid_mapped_ipv6_conversion["::ffff:13.1.68.3"] = "13.1.68.3"
	valid_mapped_ipv6_conversion["0:0:0:0:0:ffff:8190:3426"] = "129.144.52.38"
	valid_mapped_ipv6_conversion["::ffff:8190:3426"] = "129.144.52.38"
	return IPv6MappedTest{
		ip:                           *Ipv6MappedNew("::172.16.10.1").Unwrap(),
		s:                            "::ffff:172.16.10.1",
		sstr:                         "::ffff:172.16.10.1/32",
		_str:                         "0000:0000:0000:0000:0000:ffff:ac10:0a01/128",
		u128:                         str2Int("281473568475649", 10),
		address:                      "::ffff:ac10:a01/128",
		valid_mapped:                 valid_mapped,
		valid_mapped_ipv6:            valid_mapped_ipv6,
		valid_mapped_ipv6_conversion: valid_mapped_ipv6_conversion,
	}
}

func TestIpv6Mapped(tx *testing.T) {
	t := MyTesting{tx}
	t.Run("TestIpv6Mapped", func(t *MyTesting) {
		t.Run("test_initialize", func(t *MyTesting) {
			s := ipv6MappedSetup()
			t.assert_bool(true, Parse("::172.16.10.1").IsOk())
			for ip, u128 := range s.valid_mapped {
				// fmt.Printf("-%s--%s\n", ip, u128);
				if Parse(ip).IsErr() {
					fmt.Printf("%s\n", Parse(ip).UnwrapErr())
				}
				t.assert_bool(true, Parse(ip).IsOk())
				t.assert_bigint(u128, Parse(ip).Unwrap().Host_address)
			}
			for ip, u128 := range s.valid_mapped_ipv6 {
				// fmt.Printf("====%s==%s", ip, u128);
				t.assert_bool(true, Parse(ip).IsOk())
				t.assert_bigint(u128, Parse(ip).Unwrap().Host_address)
			}
		})
		t.Run("test_mapped_from_ipv6_conversion", func(t *MyTesting) {
			for ip6, ip4 := range ipv6MappedSetup().valid_mapped_ipv6_conversion {
				// fmt.Printf("+%s--%s", ip6, ip4);
				t.assert_string(ip4, Parse(ip6).Unwrap().Mapped.To_s())
			}
		})
		t.Run("test_attributes", func(t *MyTesting) {
			s := ipv6MappedSetup()
			t.assert_string(s.address, s.ip.To_string())
			t.assert_uint8(128, s.ip.Prefix.Num)
			t.assert_string(s.s, s.ip.To_s_mapped())
			t.assert_string(s.sstr, s.ip.To_string_mapped())
			t.assert_string(s._str, s.ip.To_string_uncompressed())
			t.assert_bigint(s.u128, s.ip.Host_address)
		})
		t.Run("test_method_ipv6", func(t *MyTesting) {
			s := ipv6MappedSetup().ip
			t.assert(s.Is_ipv6())
		})
		t.Run("test_mapped", func(t *MyTesting) {
			s := ipv6MappedSetup().ip
			t.assert(s.Is_mapped())
		})
	})
}

File: dart/ipaddress/test/Ipv6Mapped_test.dart
import 'package:test/test.dart';

import '../IPAddress.dart';
import '../Ipv6Mapped.dart';

class IPv6MappedTest {
  IPAddress ip;
  String s;
  String sstr;
  String string;
  BigInt u128;
  String address;
  Map<String, BigInt> valid_mapped = Map<String, BigInt>();
  Map<String, BigInt> valid_mapped_ipv6 = Map<String, BigInt>();
  Map<String, String> valid_mapped_ipv6_conversion = Map<String, String>();
  IPv6MappedTest(
      this.ip, this.s, this.sstr, this.string, this.u128, this.address);
}

IPv6MappedTest setup() {
  final ret = IPv6MappedTest(
      Ipv6Mapped.create("::172.16.10.1").value,
      "::ffff:172.16.10.1",
      "::ffff:172.16.10.1/32",
      "0000:0000:0000:0000:0000:ffff:ac10:0a01/128",
      BigInt.parse("281473568475649"),
      "::ffff:ac10:a01/128");
  ret.valid_mapped["::13.1.68.3"] = BigInt.parse(("281470899930115"));
  ret.valid_mapped["0:0:0:0:0:ffff:129.144.52.38"] =
      BigInt.parse(("281472855454758"));
  ret.valid_mapped["::ffff:129.144.52.38"] = BigInt.parse(("281472855454758"));
  ret.valid_mapped_ipv6["::ffff:13.1.68.3"] = BigInt.parse(("281470899930115"));
  ret.valid_mapped_ipv6["0:0:0:0:0:ffff:8190:3426"] =
      BigInt.parse(("281472855454758"));
  ret.valid_mapped_ipv6["::ffff:8190:3426"] = BigInt.parse(("281472855454758"));
  ret.valid_mapped_ipv6_conversion["::ffff:13.1.68.3"] = "13.1.68.3";
  ret.valid_mapped_ipv6_conversion["0:0:0:0:0:ffff:8190:3426"] =
      "129.144.52.38";
  ret.valid_mapped_ipv6_conversion["::ffff:8190:3426"] = "129.144.52.38";
  return ret;
}

void main() {
  test("test_initialize", () {
    final s = setup();
    expect(true, IPAddress.parse("::172.16.10.1").isSuccess);
    s.valid_mapped.forEach((ip, u128) {
      //println!("-{}--{}", ip, u128);
      //if IPAddress.parse(ip).is_err() {
      //    println!("{}", IPAddress.parse(ip).unwrapErr());
      //}
      expect(true, IPAddress.parse(ip).isSuccess);
      expect(u128, IPAddress.parse(ip).value.host_address);
    });
    s.valid_mapped_ipv6.forEach((ip, u128) {
      //println!("===={}=={:x}", ip, u128);
      expect(true, IPAddress.parse(ip).isSuccess);
      expect(u128, IPAddress.parse(ip).value.host_address);
    });
  });

  test("test_mapped_from_ipv6_conversion", () {
    setup().valid_mapped_ipv6_conversion.forEach((ip6, ip4) {
      //println!("+{}--{}", ip6, ip4);
      expect(ip4, IPAddress.parse(ip6).value.mapped?.to_s());
    });
  });

  test("test_attributes", () {
    final s = setup();
    expect(s.address, s.ip.to_string());
    expect(128, s.ip.prefix.num);
    expect(s.s, s.ip.to_s_mapped());
    expect(s.sstr, s.ip.to_string_mapped());
    expect(s.string, s.ip.to_string_uncompressed());
    expect(s.u128, s.ip.host_address);
  });

  test("test_method_ipv6", () {
    expect(setup().ip.is_ipv6(), true);
  });

  test("test_mapped", () {
    expect(setup().ip.is_mapped(), true);
  });
}
